exp/addons/controllers/clusterresourceset_helpers.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"github.com/pkg/errors"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	addonsv1 "sigs.k8s.io/cluster-api/exp/addons/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util"
+	utilresource "sigs.k8s.io/cluster-api/util/resource"
+	utilyaml "sigs.k8s.io/cluster-api/util/yaml"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func apply(ctx context.Context, c client.Client, data []byte) error {
+	// Transform the yaml in a list of objects, so following transformation can work on typed objects (instead of working on a string/slice of bytes)
+	objs, err := utilyaml.ToUnstructured(data)
+	if err != nil {
+		return errors.Wrapf(err, "failed converting data to unstructured objects")
+	}
+
+	sortedObjs := utilresource.SortForCreate(objs)
+	for i := range sortedObjs {
+		err = applyUnstructured(ctx, c, &objs[i])
+		if err != nil {
+			return errors.Wrapf(err,
```

Comment:
Let's try to apply as much as we can, aggregate any/all errors, and wait to return the aggregate at the end?